components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FirefoxAccount.kt

```diff
@@ -0,0 +1,283 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.appservices.fxaclient
+
+import android.util.Log
+import com.sun.jna.Pointer
+import mozilla.appservices.fxaclient.rust.FxaHandle
+import mozilla.appservices.fxaclient.rust.LibFxAFFI
+import mozilla.appservices.fxaclient.rust.RustError
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * FirefoxAccount represents the authentication state of a client.
+ */
+class FirefoxAccount(handle: FxaHandle, persistCallback: PersistCallback?) : AutoCloseable {
+    private var handle: AtomicLong = AtomicLong(handle)
+    private var persistCallback: PersistCallback? = persistCallback
+
+    /**
+     * Create a FirefoxAccount using the given config.
+     *
+     * This does not make network requests, and can be used on the main thread.
+     */
+    constructor(config: Config, persistCallback: PersistCallback? = null)
+    : this(rustCall { e ->
+        LibFxAFFI.INSTANCE.fxa_new(config.contentUrl, config.clientId, config.redirectUri, e)
+    }, persistCallback) {
+        // Persist the newly created instance state.
+        this.tryPersistState()
+    }
+
+    companion object {
+        /**
+         * Restores the account's authentication state from a JSON string produced by
+         * [FirefoxAccount.toJSONString].
+         *
+         * This does not make network requests, and can be used on the main thread.
+         *
+         * @return [FirefoxAccount] representing the authentication state
+         */
+        fun fromJSONString(json: String, persistCallback: PersistCallback? = null): FirefoxAccount {
+            return FirefoxAccount(rustCall { e ->
+                LibFxAFFI.INSTANCE.fxa_from_json(json, e)
+            }, persistCallback)
+        }
+    }
+
+    interface PersistCallback {
+        fun persist(data: String)
+    }
+
+    /**
+     * Registers a PersistCallback that will be called every time the
+     * FirefoxAccount internal state has mutated.
+     * The FirefoxAccount instance can be later restored using the
+     * `fromJSONString` class method.
+     * It is the responsibility of the consumer to ensure the persisted data
+     * is saved in a secure location, as it can contain Sync Keys and
+     * OAuth tokens.
+     */
+    fun registerPersistCallback(persistCallback: PersistCallback) {
+        this.persistCallback = persistCallback
+    }
+
+    /**
+     * Unregisters any previously registered PersistCallback.
+     */
+    fun unregisterPersistCallback() {
+        this.persistCallback = null
+    }
+
+    private fun tryPersistState() {
+        this.persistCallback?.let {
+            val json: String
+            try {
+                json = this.toJSONString()
+            } catch (e: FxaException) {
+                Log.e("FirefoxAccount", "Error serializing the FirefoxAccount state.")
```

Comment:
Can consumers do anything if this keeps failing for them? We can keep an eye on this and see if this becomes a problem in the future and deal with it then.